### PR TITLE
Series/2.x scala3

### PR DESCRIPTION
**Pull request by googley42**

### CODE REVIEW

File: dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala [hunk 1]
```diff
@@ -195,9 +196,14 @@ private[dynamodb] final case class DynamoDBExecutorImpl private (dynamoDb: Dynam
 
   private def executeBatchWriteItem(batchWriteItem: BatchWriteItem): ZIO[Any, Throwable, BatchWriteItem.Response] =
     if (batchWriteItem.requestItems.isEmpty) ZIO.succeed(BatchWriteItem.Response(None))
-    else
+    else {
+      // need to explicitly type this for Scala 3 compiler
+      val t: zio.Schedule[Any, Throwable, Any] = batchWriteItem.retryPolicy.whileInput {
```
Comment (ioleo) [inline dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala:201]: Perhaps we should find more meaningful names, like `retrySchedule`

File: dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala [hunk 2]
```diff
@@ -300,51 +304,52 @@ private[dynamodb] final case class DynamoDBExecutorImpl private (dynamoDb: Dynam
   private def executeBatchGetItem(batchGetItem: BatchGetItem): ZIO[Any, Throwable, BatchGetItem.Response] =
     if (batchGetItem.requestItems.isEmpty)
       ZIO.succeed(BatchGetItem.Response())
-    else
+    else {
+      val t: zio.Schedule[Any, Throwable, Any] = batchGetItem.retryPolicy.whileInput {
```
Comment (ioleo) [inline dynamodb/src/main/scala/zio/dynamodb/DynamoDBExecutorImpl.scala:308]: Same

File: dynamodb/src/test/scala/zio/dynamodb/AttributeValueRoundTripSerialisationSpec.scala [hunk 3]
```diff
@@ -72,14 +74,30 @@ object AttributeValueRoundTripSerialisationSpec extends ZIOSpecDefault {
   ): Serializable =
     Serializable(Gen.mapOf(Gen.string, genV), ToAttributeValue[Map[String, V]], FromAttributeValue[Map[String, V]])
 
-  private def serializableList[V: ToAttributeValue: FromAttributeValue](
+  private def serializableList[V: ToAttributeValue: FromAttributeValue: Schema](
     genV: Gen[Sized, V]
-  ): Serializable =
+  ): Serializable = {
+    val _ = implicitly[ToAttributeValue[V]]
```
Comment (ioleo) [inline dynamodb/src/test/scala/zio/dynamodb/AttributeValueRoundTripSerialisationSpec.scala]: Can we replaced with `@scala.annotation.nowarn("msg=never used")` ?

File: dynamodb/src/test/scala/zio/dynamodb/AttributeValueRoundTripSerialisationSpec.scala [hunk 4]
```diff
@@ -72,14 +74,30 @@ object AttributeValueRoundTripSerialisationSpec extends ZIOSpecDefault {
...
     Serializable(Gen.listOf(genV), ToAttributeValue[Iterable[V]], FromAttributeValue[Iterable[V]])
+  }
 
   private val serializableStringSet: Serializable =
     Serializable(Gen.setOf(Gen.string), ToAttributeValue[Set[String]], FromAttributeValue[Set[String]])
 
+  implicit def iterableSchema[A: Schema, Col[A] <: Iterable[A]]: Schema[Col[A]] =
```
Comment (ioleo) [inline dynamodb/src/test/scala/zio/dynamodb/AttributeValueRoundTripSerialisationSpec.scala]: We should move these implicits outside of spec, to `src/main`, so they are available elsewhere?

File: dynamodb/src/test/scala/zio/dynamodb/TransactionModelSpec.scala [hunk 5]
```diff
@@ -123,7 +123,7 @@ object TransactionModelSpec extends ZIOSpecDefault {
       hasField(
         "invalidActions",
         a => {
-          val b: Iterable[DynamoDBQuery[Any, Any]] = a.invalidActions.toIterable
+          val b: Iterable[DynamoDBQuery[Any, Any]] = a.invalidActions.toList
```
Comment (ioleo) [inline dynamodb/src/test/scala/zio/dynamodb/TransactionModelSpec.scala:126]: Can be replaced with `asInstanceOf` one-liner?

File: examples/src/main/scala/zio/dynamodb/examples/UpdateExpressionExamples.scala [hunk 6]
```diff
@@ -17,18 +18,18 @@ object UpdateExpressionExamples extends App {
...
-    )
+  val expr1              = $("one[2]").set(1)
+  val expr2              = expr1 + $("one[2]").set($("two"))
+  val expr3              = expr2 + $("one[2]").setIfNotExists("v2")
+  val expr4              = expr3 + $("one[2]").setIfNotExists($("two"), "v2")
+  val expr5              = expr4 + $("one[2]").appendList(List("x1", "x2"))
+  val expr6              = expr5 + $("one[2]").prependList(List("x", "x2"))
+  val expr7              = expr6 + $("one[2]").add(1)
+  // scala 3 needs type hint
+  val expr8: Action[Any] = expr7 + $("one[2]").remove.asInstanceOf[RenderableAction[Any]]
```
Comment (ioleo) [inline examples/src/main/scala/zio/dynamodb/examples/UpdateExpressionExamples.scala:29]: Perhaps `asInstanceOf` is enough?

File: examples/src/main/scala/zio/dynamodb/examples/UpdateItemExamples.scala [hunk 7]
```diff
@@ -2,16 +2,20 @@ package zio.dynamodb.examples
...
 
   val pi1: DynamoDBQuery[Nothing, Option[Item]] = updateItem("tableName1", PrimaryKey("field1" -> 1)) {
-    $("foo.bar").set("a_value") +
-      $("bar.foo").remove +
-      $("foo.foo").appendList(Chunk("s")) +
-      $("foo.foo").prependList(Chunk("s")) +
-      $("baz.fooSet").deleteFromSet(Set(Set("el1")))
+    val expr1              = $("foo.bar").set("a_value")
+    // scala 3 needs type hint
+    val expr2: Action[Any] = expr1 + $("bar.foo").remove.asInstanceOf[RenderableAction[Any]]
```
Comment (ioleo) [inline examples/src/main/scala/zio/dynamodb/examples/UpdateItemExamples.scala:14]: Same

**Review by ioleo — COMMENTED** [review]
A few remarks to consider once all tests are passing.

**Comment by googley42** [comment]
So the ToAttributeValue implicit errors have been fixed. Also as a sanity check I have created this new branch/PR that aligns the Scala matrix with zio-schema - as we have a deep dependency on it, also we can use it as a reference point for cross compilation issues https://github.com/zio/zio-dynamodb/pull/188 . Both Scala2.12 and Scala2.13 builds pass in this new PR, leaving only Scala 3 issues

**Comment by googley42** [comment]
superseded by https://github.com/zio/zio-dynamodb/pull/188

**Pull request closed without merging**